### ISSUE TITLE
Bring derived columns to their parent data groups 

### ISSUE THREAD

**Issue by janagombitova** [akvo/akvo-lumen#2740]
# Context
With supporting RQG data in Lumen (https://github.com/akvo/akvo-lumen/issues/2127) we changed how we show data in the data grid. We added a navigation option via the data groups (based on the form question groups) (https://github.com/akvo/akvo-lumen/issues/2705). 

The new UI questions how we handle derived columns. Derived columns are **new columns created based on a transformation action using data from one or multiple existing columns**. 

## Why do we have this issue? What are we trying to solve? Current status quo
In the past, once Salim derives new column we place it at the end of the data grid into a new group called `Transformations`. Salim has to scroll a lot. If he wants to see the new column, he might scroll horizontally through hundreds of columns. If he wanted to check the original column data and the transformed data, he needed to do a lot of back and forth scrolling. This is addressed by the new data groups navigation. 

<img width="1239" alt="Screen Shot 2020-06-29 at 10 05 27" src="https://user-images.githubusercontent.com/12456965/85999734-e2765080-ba0c-11ea-9d55-15cbbff73474.png">

So in the new UI, derived columns land in the `Transformations` data group. But how to handle RQG data based derived columns? Because the number of rows does not match, we thought of having all derived non-RQG data columns in the `Transformations` group and the RQG derived columns in the RQG data group (because the number of rows doesn’t match), creating inconsistency on how we handle derived columns, as some are in a specific group and some in the generic one. 

So the idea is to have each derived column live in a data group based on its parent column. 

## How will this benefit the users?
* we will create more consistency in how we handle elements. So if something behaves one way in one situation, it behaves the same way in all situations 
* bring the derived data closer to the original data column, allowing easier checking and creating an easier overview of the data

## How will this benefit Akvo? 
* by keeping on simplifying how we allow uses to work with data, we strengthen also this side of the product. Besides having the best app, we then also have the simplest tool on working with large datasets enhancing our market position
* users like simplicity when it comes to data, this can provide it 

# Opportunity 

## The idea 
The idea is that every time Salim creates a new derived column, we show it in the parent data group, the data group in which the original column is located. 

### To consider 
* How might we show that a data group has derived columns? 
* How might we indicate that a column is a derived one once in a data group (icon, highlight, colour, etc)? 
* How will we handle transformations what use more than one column, thus the columns can come from different data groups? 
    * Combine columns 
    * Derive column - javascript 
    * Generate geopoints 
* How might we handle columns merged in from a different dataset? 
* Anything else? 

Extra idea/thought just to make sure we consider all alternatives: We will start by bringing the column to the end of the group. The things to consider have to be handled regardless of the following point. --> Is there a technical reason in place, why not bring it after the original column? Anything to do with the update logic? @iperdomo @tangrammer 

### Backwards compatibility 
With **existing** datasets that have derived columns: 
* For datasets coming from Flow: 
    * will we match the derived columns to the data groups? 
    * will this show, once Salim updates a dataset? So if he does not update, nothing changes? (This is ok. I just want to make sure we are clear on the expectations.) 
* For nonFlow datasets - nothing changes? 
* Anything else?

# Next steps

**Comment by Kiarii** [comment]
indicating derived cols in their parent datagroups may be achieved by adding an [fx] flag next to the col header 
- this enables us to maintain some visual clarity on the grid, while making it easy for the user to id derived cols when they scan the col headers.. 
- the [fx] flag / badge precedes the col header since atm we are not gracefully supporting full col-header names; i.e. for long col-header names the flag would be out of view. When we achieve a graceful presentation of long col header names we might rethink the placement of the [fx] flag therein
<img width="1301" alt="Screenshot 2020-07-06 at 13 02 59" src="https://user-images.githubusercontent.com/12692444/86581866-6094a300-bf89-11ea-8803-402c02a2ae7f.png">

I think we should keep open the idea of creating an extra data group for derived cols where the user can have a quick handle to all their derived cols on the viewing and even more so on the data analysis side of things..

**Comment by janagombitova** [comment]
@Kiarii looks good. 

Second question: What about the indication in the data group navigation? We said that there it might also be good to show that the group has derived columns.

**Comment by Kiarii** [comment]
(in contemplating this ..) consider a scenario that Salim has derived cols in all data groups, in which case there would be an [fx] flag on each of the data groups but this would not really help him find a derived col of interest.. he would still have to go from group to group in search of a col of interest
I am thus inclined to think that grouping derived cols - in some way (not necessarily as a data group) - might be the next step in the iteration.. perhaps we could take lessons from releasing the first step.. on the analysis it makes even more sense to have those grouped, then they are easy to find

**Comment by janagombitova** [comment]
I am very much in favor of an iterative approach. Ok. Let’s:
- Move the derived columns to their groups
- Indicate with the [fx] flag on column level
- Learn of an indication is needed on group level
- Learn how one identifies and users the derived columns when analyzing (in visualizations)

@marvinkome  could we implement the indication of a derived column already now (as a first step) and then work on moving them to the patent groups?

**Comment by marvinkome** [comment]
From the data being sent from the backend, there's no way to know if a column is a derived column

**Comment by janagombitova** [comment]
Ok. Then we wait for the entire change for once the backend set up is in place

**Comment by iperdomo** [comment]
After a conversation with @janagombitova @Kiarii @tangrammer @kardan @muloem we have decided:

* *Derived* columns to their *parent* group, meaning no extra *Transformation* data group  for now
* We're going to limit the transformations that use more than one (1) column to be columns from the same group
* We're going to gather data related to the use of transformations across different groups, e.g. when combining columns, or JavaScript expressions
  * Based on the collected data on transformations decide if we move forward with the decision of limiting transformations to one group, or revisit the limitation

**Comment by Kiarii** [comment]
NB: another rationale for 'grouping' derived column was to make it easier to find these when building visualizations; assuming that they are harder to locate when they are spread throughout the data groups. In addition, 'grouping' derived cols for the user should not necessarily warrant special handling on the backend - imho - on the user side, the grouping is a way to view cols with a 'derived' "flag" easily..

**Comment by janagombitova** [comment]
Please let's not forget to add the UI icon to a derived column, so users know this is a new derived column in the data group (on the column level).

**Comment by janagombitova** [comment]
Tested on demo.akvolumen and all works really well. The column is added to the end of the data group and cross data group transformations are not allowed at the moment. So all works as expected. 

But the user, has no indication on where the new column is now being added. So @kardan can we add the icon - as Kiarii designed in one of the comments above - to the transformed columns? This way Salim has an idea on what is a transformed column.

**Comment by janagombitova** [comment]
@tangrammer this is another issue to close off for the `data groups` work. What is remaining is to add the `fx` icon to the transformed column (see: https://app.zenhub.com/workspaces/akvo-k2-55d493e09e846e9104ab9478/issues/akvo/akvo-lumen/2740#issuecomment-654142724) Could you handle this please?
